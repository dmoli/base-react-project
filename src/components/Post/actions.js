import api from '../../api';

export const actionTypes = {
  SET_POST: 'SET_POST',
  SHOW_ERROR: 'SHOW_ERROR',
};

export const setPost = function setPost(posts) {
  return {
    type: actionTypes.SET_POST,
    payload: posts,
  };
};

const nextPageFailure = function nextPageFailure() {
  return {
    type: actionTypes.SHOW_ERROR,
  };
};

export const nextPage = function nextPage() {
  return async (dispatch, getState) => {
    try {
      const state = getState();
      const currentPage = state.posts.page;
      const posts = await api.posts.getList(currentPage);
      dispatch(
        setPost(posts),
      );
      return posts;
    } catch (e) {
      return dispatch(nextPageFailure());
    }
  };
};
