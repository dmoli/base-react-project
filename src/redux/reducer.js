import { combineReducers } from 'redux';

import postReducer from '../components/Post/reducer';

const reducer = combineReducers({
  posts: postReducer,
});

export default reducer;
