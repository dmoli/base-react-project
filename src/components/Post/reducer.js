import { combineReducers } from 'redux';

import { actionTypes } from './actions';
import initialState from '../../redux/state';

function errorReducer(state = initialState.posts.error, action = {}) {
  switch (action.type) {
    case actionTypes.SHOW_ERROR:
      return state;
    default:
      return state;
  }
}

function pageReducer(state = initialState.posts.page, action = {}) {
  switch (action.type) {
    case actionTypes.SET_POST:
      return state + 1;
    default:
      return state;
  }
}

function entitiesReducer(state = initialState.posts.entities, action = {}) {
  switch (action.type) {
    case actionTypes.SET_POST:
      return state.concat(action.payload);
    default:
      return state;
  }
}

const postReducer = combineReducers({
  error: errorReducer,
  page: pageReducer,
  entities: entitiesReducer,
});

export default postReducer;
