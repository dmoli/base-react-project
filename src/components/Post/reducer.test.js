/* eslint-env jest */
import reducer from './reducer';
import initialState from '../../redux/state';
import { setPost } from './actions';

const items = [
  {
    id: 1,
    title: 'title test 1',
  },
  {
    id: 2,
    title: 'title test 2',
  },
  {
    id: 3,
    title: 'title test 3',
  },
];

describe('Post reducer', () => {
  it('should return the initial state', () => {
    expect(reducer(undefined, {})).toEqual(initialState.posts);
  });

  it('should handle SET_POST', () => {
    // call reducer to exect an action setPost
    expect(
      reducer(undefined, setPost(items)),
    ).toMatchSnapshot();
  });
});
