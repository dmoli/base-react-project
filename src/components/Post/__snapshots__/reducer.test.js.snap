// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post reducer should handle SET_POST 1`] = `
Object {
  "entities": Array [
    Object {
      "id": 1,
      "title": "title test 1",
    },
    Object {
      "id": 2,
      "title": "title test 2",
    },
    Object {
      "id": 3,
      "title": "title test 3",
    },
  ],
  "error": false,
  "page": 2,
}
`;
